src/ml/neural_net/style_transfer/mps_style_transfer.m

```diff
@@ -0,0 +1,419 @@
+                                      labels:gramMatrixStyleLossFourthReLU.resultImage
+                              lossDescriptor:styleDesc];
+
+    MPSNNForwardLossNode *contentLossNode
+      = [MPSNNForwardLossNode nodeWithSource:_contentVgg.reluOut3
+                                      labels:_contentVggLoss.reluOut3
+                              lossDescriptor:contentDesc];
+
+     MPSNNAdditionNode* addLossStyle1Style2
+      = [MPSNNAdditionNode nodeWithSources:@[styleLossNode1.resultImage,
+                                             styleLossNode2.resultImage]];
+
+    MPSNNAdditionNode* addLossStyle3Style4
+      = [MPSNNAdditionNode nodeWithSources:@[styleLossNode3.resultImage,
+                                             styleLossNode4.resultImage]];
+
+    MPSNNAdditionNode* addTotalStyleLoss 
+      = [MPSNNAdditionNode nodeWithSources:@[addLossStyle1Style2.resultImage,
+                                             addLossStyle3Style4.resultImage]];
+
+    MPSNNAdditionNode* totalLoss
+      = [MPSNNAdditionNode nodeWithSources:@[addTotalStyleLoss.resultImage,
+                                             contentLossNode.resultImage]];
+
+    totalLoss.resultImage.handle = [[TCMPSGraphNodeHandle alloc] initWithLabel:@"totalLossValue"];
+    totalLoss.resultImage.exportFromGraph = YES;
+    totalLoss.resultImage.synchronizeResource = YES;
+    totalLoss.resultImage.imageAllocator = [MPSImage defaultAllocator];
+
+    MPSNNInitialGradientNode *initialGradient
+      = [MPSNNInitialGradientNode nodeWithSource:totalLoss.resultImage];
+
+    BOOL resultsAreNeeded[] = { YES, YES };
+
+    NSArray<MPSNNFilterNode*>* lastNodes 
+      = [initialGradient trainingGraphWithSourceGradient:initialGradient.resultImage
+                                             nodeHandler: nil];
+
+    _trainingGraph = [MPSNNGraph graphWithDevice:_dev
+                                    resultImages:@[lastNodes[0].resultImage, lastNodes[1].resultImage]
+                                resultsAreNeeded:&resultsAreNeeded[0]];
+
+    _inferenceGraph = [MPSNNGraph graphWithDevice:_dev
+                                      resultImage:_model.forwardPass
+                              resultImageIsNeeded:YES];
+
+    _trainingGraph.format = MPSImageFeatureChannelFormatFloat32;
+    _inferenceGraph.format = MPSImageFeatureChannelFormatFloat32;
+  }
+  return self;
+}
+ 
+- (NSDictionary<NSString *, NSData *> *)exportWeights {
+  return [_model exportWeights];
+}
+
+- (NSDictionary<NSString *, NSData *> *)predict:(NSDictionary<NSString *, NSData *> *)inputs {
+  NSUInteger contentWidth;
+  NSUInteger contentHeight;
+  NSUInteger contentChannels;
+
+  [inputs[@"imageWidth"] getBytes:&contentWidth length:sizeof(contentWidth)];
+  [inputs[@"imageHeight"] getBytes:&contentHeight length:sizeof(contentHeight)];
+  [inputs[@"imageChannels"] getBytes:&contentChannels length:sizeof(contentChannels)];
```

Comment:
This seems pretty brittle. Does this mean that some of the `NSData` values store arrays of floats, and others are just `NSUInteger`? This is especially dangerous because on modern Macs, `float` is 32-bit and `NSUInteger` is 64-bit.  It seems safer either to pass this kind of thing around outside of these dictionaries of float arrays, or at least to encode them as floating point values.

Reply:
Refactored this out to properties.